Documentation/design-docs/raw-eventlistener.md

```diff
@@ -0,0 +1,81 @@
+# Raw EventListener API #
+
+The goal of this design document is to describe what a Raw EventListener API will look like and how it will work.
+
+A raw API is one that dispatches events, but does not decode the data payload.  It simply provides a raw blob and a metadata blob that can be used to interpret the data when desired.
+
+## API Surface ##
+
+The existing EventListener API usage pattern is to:
+
+1. Create a new class that derives from EventListener.
+2. Implement the OnEventSourceCreated method to get callbacks when new EventSource objects are created (this allows EventListeners to subscribe to events).
+3. Implement the OnEventWritten method to get callbacks whenever a subscribed event is dispatched.
+
+The proposed extension of this pattern is the following:
+
+1. Add a new EventListener constructor that takes an EventListenerSettings enum parameter:
+
+```
+[Flags]
+public enum EventListenerSettings
+{
+   None,
+   RawEventDispatch
+}
+``` 
+
+This parameter is used to specify the desired dispatch behavior (in this case, do not deserialize event payloads).
+
+2. Depending on the configuration, EventListener will call the traditional (deserialized payload) API or the raw API.
+
+The new raw dispatch API will be:
+
+```
+public void OnEventWrittenRaw(RawEventWrittenEventArgs args);
+ 
+public sealed class RawEventWrittenEventArgs
```

Comment:
In the private API pass, I'm wondering how best to approach this. We'd either need to make `RawEventWrittenEventArgs` public, or pass it as an `object` and reflect out these properties (or, more likely, emit some code at runtime to handle it). While we're not terribly concerned about perf in the initial passes, that's a pretty significant hit to take.

For the private API, and if the Metadata/Payload blobs contain the bulk of the data in EventWrittenEventArgs, we could get around some of the complexity by simply passing the component parts as arguments, since they are all public types already. So the callback could be something like

```
void OnEventWrittenRaw(EventSource eventSource, int eventId, ReadOnlySpan<byte> metadata, ReadOnlySpan<byte> payload)
```

Reply:
I think it would be reasonable to have the private API pass the arguments directly.  My goal with the args class is to avoid the ever-increasing overloads problem.

Reply:
Agreed, I think that makes a lot of sense as a public API